File: scaling/scripts/extract_RR.R
#!/usr/bin/env Rscript

# Author : Titouan Laessle
# Copyright 2017 Titouan Laessle
# License : MIT

library(data.table)

### Infer recombination rates, using as input the coordinates files produced by the homonyme Python script.
# Will input the arguments:
# 1. path to the input file = coordinates of each of the windows of the species' genome
# 2. path to the RR spline functions of the species
# 3. path to the output table of recombination rate per windows
args <- commandArgs(trailingOnly=TRUE)

# Load the windows' coordinates
coordinates <- fread(args[1])
# The chromosome is a factor as default, which creates trouble afterward
coordinates$V1 <- as.character(coordinates$V1)

# Load the RR spline fits
load(args[2]) ; fit_spline <- fit_spline

split_coord <- split(coordinates, coordinates$V1)

# We will create a vector of recombination rate as long as there is rows/windows on the coordinates table
windows_rr <- sapply(1:length(split_coord), function(each_chrom) {
  chrom_windows <- split_coord[[each_chrom]]
  chrom_fit <- fit_spline[[which(names(fit_spline) == chrom_windows$V1[1])]]
  # The first and last markers position (bp) usually marks the boundaries of telomeres
  first_marker <- chrom_fit$x[1] ; last_marker <- chrom_fit$x[length(chrom_fit$x)]
  # We want a table at the end, here with sapply, we end up with column = each window, whereas we want the reverse (hence t())
  sapply(1:nrow(chrom_windows), function (each_window) {
    window_line <- chrom_windows[each_window,]
    # If not telomeres, compute average recombination rate
    if (window_line$V3 < last_marker && window_line$V3  > first_marker){
      middle_window <- mean(c(as.numeric(window_line$V2), as.numeric(window_line$V3)))
      predicted_rr <- predict(chrom_fit, middle_window)$y
      # Second level of control: if the recombination rate is < 0, mark as 0 instead
      if (predicted_rr < 0) {
        return(0)
      # Else, can return the recombination rate
      } else {
        return(predicted_rr)
      }
    # Else, in the telomere = 0 recombination rate
    } else {
      return(0)
    }
  })
})

all_windows_rr <- matrix(c(coordinates$V1, unlist(windows_rr)), ncol =2)

write.table(all_windows_rr, args[3], quote = FALSE, row.names = FALSE, col.names = FALSE, sep = '\t')

File: scaling/scripts/extract_RR.py
#!/usr/bin/env python3

"""Extract the percentage of windows' average recombination rate
"""
from Bio import SeqIO
import math
import sys
import os
import subprocess

__author__ = "Titouan Laessle"
__copyright__ = "Copyright 2017 Titouan Laessle"
__license__ = "MIT"

# Wanted window size:
window_size = int(sys.argv[1])
# Species sample windows path:
species_sample = str(sys.argv[2])
# Species abbreviation:
species = '_'.join(str(species_sample.split('/')[-1]).split('_')[:2])
# Temporary file containing the sample windows coordinates
species_temp = species + '_temp_all_windows'
# Species recombination rate as spline functions:
species_RR = str(sys.argv[3])
# Output path:
output = str(sys.argv[4])


###
# Check if parent directory is present, if not create it
###
def checking_parent(file_path):
    # We don't need the file name, so will take everything but the last part
    parent_directories = '/'.join(file_path.split('/')[0:(len(file_path.split('/')) - 1)])
    # As we uses parallel, we ended up we one thread doesn't seeing the directory, attempting
    # creating it, while another just did the same -> error "The file already exist", and stopped everything...
    try:
        if not os.path.exists(parent_directories):
            os.makedirs(parent_directories)
    except:
        pass


###
# Fetch a fasta file, and clean it (remove N or n, which stands for "any nucleotides)
# Note that if the fasta file contain multiple sequences, only the first will have its CGR computed !
# Input:
#   - fasta_file : Path to the file containing the sequence one wants the CGR computed on
###
def fetch_fasta(fasta_file):
    # Will only take the first sequence of the fasta file
    try:
        records = list(SeqIO.parse(fasta_file, "fasta"))
    except:
        print("Cannot open %s, check path!" % fasta_file)
        sys.exit()
    return records


###
# Write a BED file containing the coordinates (first/last) of each windows not containing any N in them
# Input:
#   - records : fetched sequence (fasta) one wants the CGR computed on
#   - window_size : size of the sliding window
#   - species_temp : path of the file which will be written on (removed at the end of the script)
# Output:
#   - A BED file, each row corresponds to a window, and on each row one can find the record ids and
#       boundaries of the window.
###
def all_coordinates_BED(records, window_size, species_temp):
    with open(species_temp, 'w') as outfile:
        for each_record in range(len(records)):
            window_id = '_'.join([records[each_record].id.split('_')[0], records[each_record].id.split('_')[1]])
            start = int(records[each_record].id.split('_')[2])
            end = start + window_size
            # If any character in the sequence is NOT a standard nucleotides (including unknown nucleotides),
            # do NOT compute:
            if not any([c not in 'ATCGatcg' for c in records[each_record].seq[start:end]]):
                to_write = '\t'.join([window_id, str(start), str(end)])
                outfile.write(to_write + '\n')


# Fetch all the records from this species fasta
samples = fetch_fasta(species_sample)

all_coordinates_BED(samples, window_size, species_temp)

# Actual averaging of the spline fit of the recombination rates, for each windows
subprocess.call(['Rscript', 'scripts/extract_RR.R', species_temp, species_RR, output])

# Will remove the temp file
os.remove(species_temp)
